File: core/src/processing/event/TouchEvent.java
/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
  Part of the Processing project - http://processing.org

  Copyright (c) 2012-16 The Processing Foundation

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation, version 2.1.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA
*/

package processing.event;

/*
http://developer.android.com/guide/topics/ui/ui-events.html
http://developer.android.com/reference/android/view/MotionEvent.html
http://android-developers.blogspot.com/2010/06/making-sense-of-multitouch.html
http://www.techrepublic.com/blog/app-builder/use-androids-gesture-detector-to-translate-a-swipe-into-an-event/1577
*/
public class TouchEvent extends Event {
  static public final int START = 1;
  static public final int END = 2;
  static public final int CANCEL = 3;
  static public final int MOVE = 4;

  protected int action;

  protected int button;
  protected int numPointers;
  protected int[] pointerId;
  protected float[] pointerX;
  protected float[] pointerY;
  protected float[] pointerSize;
  protected float[] pointerPressure;

  public TouchEvent(Object nativeObject, long millis, int action, int modifiers,
                    int button) {
    super(nativeObject, millis, action, modifiers);
    this.flavor = TOUCH;
    this.button = button;
  }

  public void setNumPointers(int n) {
    numPointers = n;
    pointerId = new int[n];
    pointerX = new float[n];
    pointerY = new float[n];
    pointerSize = new float[n];
    pointerPressure = new float[n];
 }


 public void setPointer(int idx, int id, float x, float y, float s, float p) {
   pointerId[idx] = id;
   pointerX[idx] = x;
   pointerY[idx] = y;
   pointerSize[idx] = s;
   pointerPressure[idx] = p;
 }


 public int getNumPointers() {
   return numPointers;
 }


 public PPointer getPointer(int idx) {
   PPointer pt = new PPointer();
   pt.id = pointerId[idx];
   pt.x = pointerX[idx];
   pt.y = pointerY[idx];
   pt.size = pointerSize[idx];
   pt.pressure = pointerPressure[idx];
   return pt;
 }


 public int getPointerId(int idx) {
   return pointerId[idx];
 }


 public float getPointerX(int idx) {
   return pointerX[idx];
 }


 public float getPointerY(int idx) {
   return pointerY[idx];
 }


 public float getPointerSize(int idx) {
   return pointerSize[idx];
 }


 public float getPointerPressure(int idx) {
   return pointerPressure[idx];
 }


 public int getButton() {
   return button;
 }


 public PPointer[] getTouches(PPointer[] touches) {
   if (touches == null || touches.length != numPointers) {
     touches = new PPointer[numPointers];
     for (int idx = 0; idx < numPointers; idx++) {
       touches[idx] = new PPointer();
     }
   }
   for (int idx = 0; idx < numPointers; idx++) {
     touches[idx].id = pointerId[idx];
     touches[idx].x = pointerX[idx];
     touches[idx].y = pointerY[idx];
     touches[idx].size = pointerSize[idx];
     touches[idx].pressure = pointerPressure[idx];
   }
   return touches;
 }


  public class PPointer {
    public int id;
    public float x, y;
    public float size;
    public float pressure;
  }
}

File: core/src/processing/event/TouchEvent.kt
/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
  Part of the Processing project - http://processing.org
  Copyright (c) 2012-16 The Processing Foundation
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation, version 2.1.
  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.
  You should have received a copy of the GNU Lesser General
  Public License along with this library; if not, write to the
  Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  Boston, MA  02111-1307  USA
*/

package processing.event

/**
 * @author Aditya Rana
 * IMPORTANT NOTICE: None of the methods and inner classes in TouchEvent are
 * part of the Processing API. Don't use them! They might be changed or removed
 * without notice.
*/
open class TouchEvent(nativeObject: Any?, millis: Long, action: Int, modifiers: Int,
                      button: Int) : Event(nativeObject!!, millis, action, modifiers) {
    override var action = 0

    override var native: Any
        get() = super.native
        set(value) {
            native = value
        }

    var button: Int
        protected set
    private var numPointers = 0
    private lateinit var pointerId: IntArray
    private lateinit var pointerX: FloatArray
    private lateinit var pointerY: FloatArray
    private lateinit var pointerArea: FloatArray
    private lateinit var pointerPressure: FloatArray
    fun setNumPointers(n: Int) {
        numPointers = n
        pointerId = IntArray(n)
        pointerX = FloatArray(n)
        pointerY = FloatArray(n)
        pointerArea = FloatArray(n)
        pointerPressure = FloatArray(n)
    }

    fun setPointer(idx: Int, id: Int, x: Float, y: Float, a: Float, p: Float) {
        pointerId[idx] = id
        pointerX[idx] = x
        pointerY[idx] = y
        pointerArea[idx] = a
        pointerPressure[idx] = p
    }

    fun getNumPointers(): Int {
        return numPointers
    }

    fun getPointer(idx: Int): Pointer {
        val pt = Pointer()
        pt.id = pointerId[idx]
        pt.x = pointerX[idx]
        pt.y = pointerY[idx]
        pt.area = pointerArea[idx]
        pt.pressure = pointerPressure[idx]
        return pt
    }

    fun getPointerId(idx: Int): Int {
        return pointerId[idx]
    }

    fun getPointerX(idx: Int): Float {
        return pointerX[idx]
    }

    fun getPointerY(idx: Int): Float {
        return pointerY[idx]
    }

    fun getPointerArea(idx: Int): Float {
        return pointerArea[idx]
    }

    fun getPointerPressure(idx: Int): Float {
        return pointerPressure[idx]
    }

    fun getTouches(touches: Array<Pointer?>?): Array<Pointer?> {
        var touches = touches
        if (touches == null || touches.size != numPointers) {
            touches = arrayOfNulls(numPointers)
            for (idx in 0 until numPointers) {
                touches[idx] = Pointer()
            }
        }
        for (idx in 0 until numPointers) {
            touches[idx]!!.id = pointerId[idx]
            touches[idx]!!.x = pointerX[idx]
            touches[idx]!!.y = pointerY[idx]
            touches[idx]!!.area = pointerArea[idx]
            touches[idx]!!.pressure = pointerPressure[idx]
        }
        return touches
    }

    inner class Pointer {
        var id = 0
        var x = 0f
        var y = 0f
        var area = 0f
        var pressure = 0f
    }

    companion object {
        const val START = 1
        const val END = 2
        const val CANCEL = 3
        const val MOVE = 4
    }

    // constructor or initializer block
    init {
        flavor = TOUCH
        this.button = button
    }
}